circuits/tally/src/poseidon.leo

```diff
@@ -0,0 +1,289 @@
+circuit Poseidon {
+    /*    
+    Poseidon for fields
+    Takes 2 fields as inputs and generates one output, 
+    to use in binary merkle trees
+
+    We base our implementation in the Poseidon RS implementation
+    Which itself is based in the sage reference implementation
+    https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/poseidonperm_x5_254_3.sage
+    
+    Constants generated for bls 12 377 using https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/generate_parameters_grain.sage  
+    sage generate_parameters_grain.sage 1 0 253 6 8 60 0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001
+    then transformed from hexstring array to dec array in python with 
+    [int(x,16) for x in c_hexstring]
+    [ [ int(x,16) for x in row ] for row in m_string ]
+
+    Notes for paramaters to generate constants:
+    Main sub group order / prime number hex
+    0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001
+    = b0001001...
+     -> n = 256 - 3 (zeros) = 253
+    sage generate_parameters_grain.sage 1 0 253 3 8 57 0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001
+    
+    */
+    function poseidon_2_1(a_field: field, another_field: field) -> field {
+
+        const n_rounds_f: u32 = 8;
+        const n_rounds_p: u32 = 57;
+        //t = input len + 1
+        const t: u32 = 3;
+        let state: [field;3] = [0,a_field,another_field];
+
+        for i in 0..(n_rounds_f + n_rounds_p) {
+            state = Poseidon::ark(state, i*t);
+            state =  Poseidon::sbox(n_rounds_f, n_rounds_p, state, i);
+            state = Poseidon::mix(state);
+        }
+        return state[0];
+    }
+
+    function ark(state: [field;3], it: u32 ) -> [field;3]{
+        // state[i].add_assign(&c[it + i]);
+        const c_constants: [field;195] = 
+        [
+            1278823093112533518033294982858707370905937792805078104831550399751562346033,
```

Comment:
Are we sure is it safe to have this fields in the repo? should we check it with errol?

Reply:
Yes we are

Reply:
The constants are public in every Poseidon implementation, they are not secrets

Reply:
great, i will marge it